pkg/backends/calico/client.go

```diff
@@ -811,6 +885,258 @@ func (c *client) onUpdates(updates []api.Update) {
 	c.onNewUpdates()
 }
 
+func (c *client) updateBGPConfigCache(update api.Update) {
+	key, ok := update.KVPair.Key.(model.ResourceKey)
+	if !ok {
+		log.Infof("Error getting BGP Configuration key. %#v",ok)
+	}
+
+	resourceName := key.Name
+	v3res, ok := update.KVPair.Value.(*apiv3.BGPConfiguration)
+	if !ok {
+		log.Infof("BGP Configuration is either empty or error occurred while getting BGPConfiguration value %#v", ok)
+		v3res = nil
+	}
+
+	if resourceName == globalConfigName {
+		c.getPrefixAdvertisementsKVPair(v3res, model.GlobalBGPConfigKey{})
+		c.getListenPortKVPair(v3res, model.GlobalBGPConfigKey{})
+		c.getASNumberKVPair(v3res, model.GlobalBGPConfigKey{})
+		c.getServiceExternalIPsKVPair(v3res, model.GlobalBGPConfigKey{})
+		c.getServiceClusterIPsKVPair(v3res, model.GlobalBGPConfigKey{})
+		c.getNodeToNodeMeshKVPair(v3res, model.GlobalBGPConfigKey{})
+		c.getLogSeverityKVPair(v3res, model.GlobalBGPConfigKey{})
+	} else {
+		// update node-specific BGPConfig
+		var nodeName string
+		if strings.HasPrefix(resourceName, perNodeConfigNamePrefix) {
+			nodeName = resourceName[len(perNodeConfigNamePrefix):]
+		}
+		c.getPrefixAdvertisementsKVPair(v3res, model.NodeBGPConfigKey{Nodename: nodeName})
+		c.getListenPortKVPair(v3res, model.NodeBGPConfigKey{Nodename: nodeName})
+		c.getLogSeverityKVPair(v3res, model.NodeBGPConfigKey{Nodename: nodeName})
```

Comment:
Oh....   Sorry, that is wrong, but I've just realised that overall we're doing something a bit inelegant here in the data model here: some node-specific things are configured in a per-node BGPConfiguration, but most node-specific things are configured in Node.Spec.BGP.

Prior to this work, I think the only per-node BGPConfiguration thing is loglevel (which would rarely be used), but now we're adding communities and listen port as well.  What is the overall thinking for adding those in BGPConfiguration instead of in Node.Spec.BGP?  I'm sorry for not noticing and asking about this before!

(I guess BGPConfiguration is better because it means we can use a proper resource instead of Node annotations in the KDD case; especially so for complex data like communities.  But then I wonder why we ever used Node.Spec.BGP?)

Reply:
cc @caseydavenport

Reply:
The main reason for not having them in `Node.Spec.BGP` would be that BGPConfiguration can be global.

Reply:
> But then I wonder why we ever used Node.Spec.BGP?

yeah, I am unsure about some of the history here. I am guessing it was not an especially intentional choice to have data used by the BGP component stored both in the BGPConfiguration _and_ the Node object.

But in this case, I do think BGPConfiguration is the correct place for community strings because it can be set globally across the cluster, whereas the Node object is strictly per-node. I think a lot of use-cases for community strings will involve them being the same across nodes, so makes sense to be configured globally (and IMO even disable them per-node via validation unless there is a clear use-case for it?)

Reply:
Thanks @caseydavenport , that makes sense.